File: pardump.c
#include <stdlib.h>
#include <stdio.h>
#include <arpa/inet.h>
#include "pardump.h"


partmass_t *
ntoh_particle(partmass_t *p)
{
    uint32_t *word = (uint32_t *)p ; 
    partmass_t *swapped =0 ;

    swapped = (partmass_t *)malloc(sizeof(partmass_t)) ; 
    uint32_t *swapped_word = (uint32_t *)swapped ; 

    for (int i=0; i<(sizeof(partmass_t)/sizeof(uint32_t)); i++)
    {
        *swapped_word++ = ntohl(*word++) ; 
    }

    return swapped;
}

particle_dump_header_t *
ntoh_dump(particle_dump_header_t *p)
{
    uint32_t *word = (uint32_t *)p ; 
    particle_dump_header_t *swapped =0 ;

    swapped = (particle_dump_header_t *)malloc(sizeof(particle_dump_header_t)) ; 
    uint32_t *swapped_word = (uint32_t *)swapped ; 

    for (int i=0; i<(sizeof(particle_dump_header_t)/sizeof(uint32_t)); i++)
    {
        *swapped_word++ = ntohl(*word++) ; 
    }

    return swapped;
}

partmass_t *
read_particles(FILE *fh, int n)
{
    partmass_t *particle ; 

    particle = (partmass_t *)malloc(PARTICLE_SIZE*n) ; 
    if (particle) {
        fread(particle, PARTICLE_SIZE, n, fh) ; 
    }
    return particle ; 
}

particle_dump_t *
read_particle_dump(FILE *fh)
{
    particle_dump_t *pd ; 
    particle_dump_header_t *hdr;
    unsigned int n_particles = 0 ;
    
    
    pd  = (particle_dump_t *)malloc(sizeof(particle_dump_t)) ; 
    hdr = (particle_dump_header_t *)malloc(sizeof(particle_dump_header_t)) ; 
    if (pd && hdr) { 
        pd->header = hdr ; 
        fread(hdr, sizeof(particle_dump_header_t), 1, fh) ; 

        /* get the number of particles but switch endianness */
        n_particles = ntohl(hdr->n_particles) ; 

        /* read the actual particles */
        pd->particles = read_particles(fh, n_particles) ; 
      
    }
    return pd ;
}

void 
print_particle_dump(particle_dump_t *pd)
{
    particle_dump_header_t *hdr = pd->header ; 
    printf("Number of particles: %d\n", ntohl(hdr->n_particles)) ; 
    printf("Number of pollutants: %d\n", ntohl(hdr->n_pollutants)) ; 
    printf("Time of dump: %4d-%02d-%02d %02d:%02d\n",
             ntohl(hdr->year)+2000, ntohl(hdr->month), ntohl(hdr->day),
             ntohl(hdr->hour), ntohl(hdr->minute)) ; 
}

void 
print_particle_dump_raw(particle_dump_t *pm) 
{
    particle_dump_header_t *swapped = ntoh_dump(pm->header) ; 
    uint32_t *printable = (uint32_t *)swapped ; 
    for (int i=0; i < (sizeof(particle_dump_header_t)/sizeof(uint32_t)); i++) 
    {
        float *pfloat ;
        pfloat = (float *)printable ; 
        printf("%d: %u %f\n", i, *(printable++), *pfloat) ; 
    } 
    free(swapped) ; 
}

void 
print_particle(partmass_t *pm)
{
    partmass_t *printable = ntoh_particle(pm) ; 
    printf("(%f, %f) %fm %f(mass units) %us\nPollutant: %u; Sigma-h: %f\nVel w/v: %f/%f\n", 
        printable->particle.lon, printable->particle.lat, printable->particle.height,
        printable->mass[0], printable->particle.age,
        printable->particle.pollutant, printable->particle.sigma_h,
        printable->particle.vel_w, printable->particle.vel_v);

    free(printable) ; 
}

void 
print_particle_raw(partmass_t *pm) 
{
    partmass_t *swapped = ntoh_particle(pm) ; 
    uint32_t *printable = (uint32_t *)swapped ; 
    for (int i=0; i < (sizeof(partmass_t)/sizeof(uint32_t)); i++) 
    {
        float *pfloat ;
        pfloat = (float *)printable ; 
        printf("%d: %u %f\n", i, *(printable++), *pfloat) ; 
    } 
    free(swapped) ; 
}

void
free_particle_dump(particle_dump_t *pd)
{
    if (pd) 
    {
        free(pd->header) ; 
        free(pd->particles) ; 
        free(pd) ; 
    }
}

int 
main(int ac, char **av) 
{ 
    /* usage statement */
    if (ac != 2) {
        printf("Usage: %s <pardump_file>\n", av[0]) ; 
        return -1 ; 
    }

    /* open file */
    FILE *fh = fopen(av[1], "rb") ; 
    if (!fh) { 
        printf("Could not open file: %s\n", av[1]) ; 
        return -2;
    }

    printf("Size of partmass_t in words: %d\n", sizeof(partmass_t)/4) ; 

    /* skip undocumented dummy field */
    uint32_t dummy =0 ; 
    fread(&dummy, sizeof(dummy), 1, fh) ; 

    /* read first particle dump */
    particle_dump_t *first = read_particle_dump(fh) ; 

    print_particle_dump(first) ; 
    print_particle(first->particles) ; 
    print_particle(first->particles+2) ; 

    /* free memory and close files */
    free(first) ; 
    fclose(fh) ; 

    return 0 ; 
}

File: pardump.py
from struct import pack, unpack, calcsize ; 
from datetime import datetime, timedelta ; 
from functools import partial ; 

#
# Models hysplit particles in particle dumps. 
# Reads and writes from particle dump files. 
# writes to buffer suitable for upload to PostGIS 
# database.
#
class Particle (object) : 
    _pardump_format_no_mass = "!" + "f"*6 + "i"*5

    def __init__(self) : 
        self._pollutant_mass = None
        self._lat = None
        self._lon = None
        self._height = None
        self._sigma_h = None
        self._vel_w = None
        self._vel_v = None
        self._age = None
        self._distribution = None
        self._pollutant = None
        self._meteo_grid = None
        self._sort_index = None

    @staticmethod
    def ReadPardump(file_object, num_pollutants) : 
        mass_format = "!" + "f"*num_pollutants
        buf = file_object.read(calcsize(mass_format))
        new_obj = Particle()
        new_obj._pollutant_mass = unpack(buf, mass_format)

        buf = file_object.read(calcsize(Particle._pardump_format_no_mass))
        (new_obj._lat, new_obj._lon, new_obj._height,
            new_obj._sigma_h, new_obj._vel_w, new_obj._vel_v,
            new_obj._age, new_obj._distribution, new_obj._pollutant, 
            new_obj._meteo_grid, new_obj._sort_index) = \
            unpack(Particle._pardump_format_no_mass, buf)
        return new_obj

    @property
    def empty(self) : 
        """Returns true if this object is uninitialized"""
        return self._pollutant_mass is None

    @property
    def dead(self):  
        """Returns true if this object has lat/lon==0, a flag for particle not in use."""
        return (self._lat==0) and (self._lon == 0)

    @property
    def num_pollutants(self) : 
        return len(self._pollutant_mass)

    @property
    def mass(self) : 
        """Tuple of pollutant masses. Number of pollutants per particle is set 
        by MAXDIM. Default is 1."""
        return self._pollutant_mass

    @property 
    def lat(self) : 
        """Particle/puff latitude"""
        return self._lat

    @property
    def lon(self) : 
        """particle/puff longitude"""
        return self._lon

    @property
    def height(self) : 
        """HEIGHT is in meters agl. The height is in agl even when KMSL=1 
           is set in SETUP.CFG Note that there may be some differences between 
           the value written to the pardump file and the actual value when the 
           terrain height is large. The terrain height is not passed to the 
           pardump reading or writing routines and so the height is written 
           approximating terrain height as 0m. This results in a small difference 
           which is dependent on the terrain height and internal scaling height."""
        return self._height

    @property
    def sigma_h(self) : 
        """SIGMA-H is the horizontal puff size in meters."""
        return self._sigma_h

    @property
    def vel_w(self) :
        """VEL-W is the current value for the turbulent velocity in the vertical in m/s"""
        return self._vel_w

    @property
    def vel_v(self) : 
        """VEL-V is the current value for the turbulent velocity in the horizontal in m/s.
           Currently VEL-U is not written to the pardump file and when the model is 
           initialized from a pardump file, it assumed that vel-u = vel-v."""
        return self._vel_v

    @property
    def age(self) : 
        """particle age in seconds"""
        return self._age

    @property
    def age_timedelta(self) : 
        return timedelta(seconds=self.age)

    @property
    def distribution(self) :
        return self._distribution

    @property
    def pollutant(self) : 
        """POLLUTANT is an integer, N, which corresponds to the Nth pollutant 
           species defined in the CONTROL file."""
        return self._pollutant

    @property
    def meteo_grid(self) : 
        """METEO-GRID is n integer, N, which corresponds to the Nth meteorological 
           data grid defined in the CONTROL file."""
        return self._meteo_grid

    @property
    def sort_index(self) : 
        """Each particle or puff is assigned its own unique sort-index. 
           The sort index is used within HYSPLIT for looping through all the 
           computational particles or puffs. If a computational particle is 
           removed during the simulation, the sort indices are reassigned so 
           that there are only as many sort indices as particles. Consequently, 
           the sort index for a computational particle may change during a simulation. 
           If ichem=8, then particle deletion is turned off, and the sort index will 
           refer to the same computational particle throughout the simulation. In 
           the case of puffs, the sort indices are reassigned when puffs are deleted, 
           merged or split."""
        return self._sort_index    
        
    
class ParticleDump(object) : 
    """Essentially represents a list of particles with a given number of pollutants
       at a specific time."""

    _pardump_format = "!" + "i"*7

    def __init__(self) : 
        self._year = None
        self._month = None
        self._day   = None
        self._hour  = None
        self._minute  = None
        self._maxdim  = None
        self._particles = None

    @staticmethod
    def ReadPardump(file_object) :
        new_obj = ParticleDump()
        try: 
            buf = file_object.read(calcsize(ParticleDump._pardump_format))
            (num_particles, new_obj._maxdim,
                  new_obj._year, new_obj._month, new_obj._day, 
                  new_obj._hour, new_obj._minute) = unpack(ParticleDump._pardump_format, buf)
            new_obj._year += 2000
            new_obj._particles = [] 
            for i in range(num_particles) : 
                new_obj._particles.append(Particle.ReadPardump(file_object, new_obj._maxdim))
        except : 
            if new_obj.empty: 
                new_obj = None
        return new_obj
        
    @property
    def empty(self) : 
        return self._particles is None

    @property
    def num_particles(self) : 
        """Number of particles in this dump"""
        return len(self._particles)

    @property
    def num_pollutants(self) : 
        """Number of pollutants per particle is set by MAXDIM. Default is 1."""
        return self._maxdim

    @property
    def year(self) : 
        """Returns 4 digit year. Stored as 2 digit year in PARDUMP file."""
        return self._year

    @property
    def month(self) : 
        return self._month

    @property
    def day(self) : 
        return self._day

    @property
    def hour(self) : 
        return self._hour

    @property
    def minute(self) : 
        return self._minute

    @property
    def timestamp(self) : 
        return datetime(self.year, self.month, self.day, self.hour, self.minute)

def ReadPardumpFile(fname)  :
    particle_dump_list = [] 
    with open(fname, 'rb') as pardump : 
        #discard undocumented leading 4 bytes.
        pardump.read(4)

        for particle_dump in iter(partial(ParticleDump.ReadPardump,pardump), None) : 
            particle_dump_list.append(particle_dump)

    return particle_dump_list
